Reactive RESTEasy - Project demonstrating various issues

**Describe the bug**
I have tried RESTEasy reactive on a project and came across numerous issues. I have created a replicator project which simplifies down the various issues I have found thus far:
1) Sub resource path params are not bound correctly
2) A POST which accepts a List of non-trivial Objects fails deployment
3) Binding a BigDecimal as a path param causing booting to fail with a ClassNotFoundException
4) Returning a List<String> encodes the result incorrectly
5) Binding a resource method with multiple Method annotations results in only the first one being bound
6) Calls to sub resources pass through a ContainerRequestFilter for each sub resource level
7) Calling requestContext.getUriInfo().getMatchedResources() in a ContainerResponseFilter results in an NPE  java.lang.NullPointerException: Cannot invoke "org.jboss.resteasy.reactive.server.mapping.RuntimeResource.getClassPath()" because "this.target" is null

**Expected behavior**
All of these should work. The sample project passes all tests using standard RESTEasy

**Actual behavior**
All 7 test cases fail

**To Reproduce**
git clone https://github.com/bcluap/quarkus-examples.git
cd quarkus-examples/resteasy-reactive
mvn clean test

**Configuration**
Nothing

**Screenshots**
Run the reproducer to see failures. Change the pom to non-reactive mode and all tests pass

**Environment (please complete the following information):**
This is not environment specific

**Additional context**
NA

I've updated the reproducer with a new issue:
8) If a filter aborts with a payload > 30463 bytes and there is a WriterInterceptor then an exception occurs:
Caused by: java.lang.UnsupportedOperationException: direct buffer
	at io.netty.buffer.PooledUnsafeDirectByteBuf.array(PooledUnsafeDirectByteBuf.java:226)
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.write(ResteasyReactiveOutputStream.java:90)

This has something to do with an overflow buffer trying to use the array method on Netty PooledUnsafeDirectByteBuf which is not supported.

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/resource/basic/AbortFilterWithLargePayloadTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/resource/basic/AbortFilterWithLargePayloadTest.java
@@ -1,0 +1,73 @@
+package io.quarkus.resteasy.reactive.server.test.resource.basic;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.container.PreMatching;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.Provider;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class AbortFilterWithLargePayloadTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest testExtension = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    JavaArchive war = ShrinkWrap.create(JavaArchive.class);
+                    war.addClasses(TestResource.class, AbortingFilter.class);
+                    return war;
+                }
+            });
+
+    @Test
+    public void test() {
+        RestAssured.get("/test")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("hello"));
+
+        io.restassured.response.Response response = RestAssured.with().header("abort", "true").get("/test")
+                .then()
+                .statusCode(999).extract().response();
+        Assertions.assertEquals(30464, response.body().asByteArray().length);
+    }
+
+    @Path("/test")
+    public static class TestResource {
+
+        @GET
+        public String hello() {
+            return "hello";
+        }
+    }
+
+    @Provider
+    @PreMatching
+    public static class AbortingFilter implements ContainerRequestFilter {
+
+        @Override
+        public void filter(ContainerRequestContext requestContext) throws IOException {
+            if (requestContext.getHeaderString("abort") != null) {
+                // the magic number of 30464 is where the overflow code path used to fail
+                requestContext.abortWith(
+                        Response.status(999).type(MediaType.APPLICATION_OCTET_STREAM).entity(new byte[30464]).build());
+            }
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/ResteasyReactiveOutputStream.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/ResteasyReactiveOutputStream.java
@@ -87,8 +87,11 @@ public class ResteasyReactiveOutputStream extends OutputStream {
                     if (overflow == null) {
                         overflow = new ByteArrayOutputStream();
                     }
-                    overflow.write(data.array(), data.arrayOffset() + data.readerIndex(),
-                            data.arrayOffset() + data.writerIndex());
+                    if (data.hasArray()) {
+                        overflow.write(data.array(), data.arrayOffset() + data.readerIndex(), data.readableBytes());
+                    } else {
+                        data.getBytes(data.readerIndex(), overflow, data.readableBytes());
+                    }
                     if (last) {
                         closed = true;
                     }

